### PR TITLE
Add Last modified field to Bucket Metadata

### LINKED ISSUE #661: Add "last modified" to Metadata
**Is your feature request related to a problem? Please describe.**
Metadata was created to hold schema and allow us to easily add new fields to all objects as needed. When reflecting on the issue of versioning (like electorate), it was discussed whether to make this a general field, that is only used by the `VersionedBucket`, but accessible on all objects. Another proposal from @alpe was to add a "last modified block height" or "last modified timestamp (from header)" instead.

Reflecting on that, I think a "last modified height" would be valuable, not just for versioning, but also in general for more powerful queries in the abci app itself (or externally, but with merkle proofs). We can quickly prove the last point in time our account balance was modified, even if the tx tags were not indexed on the node in question. We could also use these to help as a heuristic in cleanup/state pruning/batch migrations/etc.

**Describe the solution you'd like**

I would like to add a field to schema: `uint64 last_modified = 2`, where `last_modified = height << 16 + index`.

Note that there may be multiple transactions in the same block that modify one item, so to ensure this number changes with each write (and allow it to be used in place of a version id field), we must include index as well. I believe 65k (2^16) is more transactions than would fit in any block, but we could make this a bit higher if needed. Currently this leaves space for 2^48 (around 400 trillion) blocks before rolling over, which could probably be trimmed down a bit safely.

In any case, I think a clever combination of block height and tx index into one uint64 would be a nice counter. And we can expose this in the Context to make it easy to uniformly calculate. This number would be present on all object and update with each save. The difference of the VersionedBucket would be that it would use this number to make the id of the object as well and create new objects on each save rather than overwriting them. 

**Describe alternatives you've considered**

Just leave Versioning and Metadata/Schema alone. They are plenty of innovation recently. Maybe we should let them settle for a while. Then again, breaking changes are always better before mainnet and before audit.

**Additional context**

Since this is a binary breaking change for all models of all extensions, rather than provide migration paths for them all, I would prefer to merge this idea (if done) before we make a semi-permanent public testnet.

### PR DESCRIPTION
**Pull request by ethanfrey**
Closes #661 
Replaces #752

After some discussions with Alex, I agree that this is the cleanest solution:

` // 1. Set the height in weave.KVStore when we cachewrap?`

I add a `GetHeight() int64` method to all KVStore, and properly set it from the last commit height in the live app, and to a provided value in all test constructors.

Add this functionality of setting it to the Migratable bucket as that is handling Metadata right now.

Updated all tests to include LastModified as part of the expected responses. Lots of tests.

Only broken code in x/namecoin and cmd/bcpd. 
But I propose to remove them in #791 and #792.
If acceptable, I will rebase on that merge and game over.
Otherwise, I will fix up legacy code :(

### CODE REVIEW

**Comment by iov-bot-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please include a CHANGELOG entry. <br />
You can find it at <a href="https://github.com/iov-one/weave/blob/master/CHANGELOG.md">CHANGELOG.md</a>.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: x/aswap/handler_test.go [hunk 1]
```diff
@@ -373,7 +373,7 @@ func TestReturnHandler(t *testing.T) {
 }
 
 func setBalance(t *testing.T, db weave.KVStore, addr weave.Address, coins coin.Coins) {
-	acct, err := cash.WalletWith(addr, coins...)
+	acct, err := cash.WalletWith(179, addr, coins...)
```
[the remark this answers is not in the document]
Comment (alpe) [reply x/aswap/handler_test.go]: sorry, it is correct here.

File: x/cash/init.go [hunk 2]
```diff
@@ -31,7 +31,7 @@ func (Initializer) FromGenesis(opts weave.Options, params weave.GenesisParams, k
 		if err := acct.Address.Validate(); err != nil {
 			return err
 		}
-		wallet, err := WalletWith(acct.Address, acct.Set.Coins...)
+		wallet, err := WalletWith(179, acct.Address, acct.Set.Coins...)
```
Comment (alpe) [inline x/cash/init.go]: Q: why is it `179` this does not look correct.

File: app/helpers.go [hunk 3]
```diff
@@ -23,6 +24,11 @@ func NewABCIStore(app abci.Application) *ABCIStore {
 	return &ABCIStore{app: app}
 }
 
+// GetHeight returns the last committed block of the application
+func (a *ABCIStore) GetHeight() int64 {
```
[the remark this answers is not in the document]
Comment (husio) [reply app/helpers.go]: It is idiomatic [not to prefix a getter with `Get`](https://golang.org/doc/effective_go.html#Getters). Applying this convention results in `LastCommittedBlockHeight` or `LastBlockHeight`.

**Comment by alpe** [comment]
I had a chance to look a bit deeper at the iavl implementation now. The implementation looks good and solves the problem to pass height all the way down to the migration bucket.
Although I was in favour of hiding this functionality for the programmers and provide a simple change without touching all methods, I feel more convinced by @husio note now:
> I think that if we had a context instance passed to every bucket method, we would not add block height information to the store.

I do agree now that changing the bucket signature to pass the context on save feels like the only proper solution that addresses the issue in a transparent and clean way.

File: x/cash/model.go [hunk 4]
```diff
@@ -154,7 +154,7 @@ type WalletBucket interface {
 // panics on error (meant as a sanity check in init)
 func ValidateWalletBucket(bucket WalletBucket) {
 	// runtime type-check the bucket....
-	db := store.MemStore()
+	db := store.MemStore(179)
```
[the remark this answers is not in the document]
Comment (ethanfrey) [reply x/cash/model.go]: Can be anything, just we need to pass a height, and verify below it was set to same value.

File: x/cash/model.go [hunk 5]
```diff
@@ -100,15 +100,15 @@ func Concat(cng Coinage, coins coin.Coins) error {
 // NewWallet creates an empty wallet with this address serves as an object for
 // the bucket.
 // NewWallet wraps Set into an object for the Bucket.
-func NewWallet(key weave.Address) orm.Object {
+func NewWallet(height int64, key weave.Address) orm.Object {
 	return orm.NewSimpleObj(key, &Set{
-		Metadata: &weave.Metadata{Schema: 1},
+		Metadata: &weave.Metadata{Schema: 1, LastModified: height},
```
[the remark this answers is not in the document]
Comment (ethanfrey) [reply x/cash/model.go]: It is. This creates the models for the expectations.

This is not used in the code that creates the wallets, but rather in test fixtures

**Comment by ethanfrey** [comment]
Closing this issue, as popular opinion is to modify the orm APIs to do this cleanly (albeit with breaking changes)

**Pull request closed without merging**